app/controllers/settings_controller.rb

```diff
@@ -127,13 +127,14 @@ def twofa_update
 
     @user.totp_secret = session[:totp_secret]
     if @user.authenticate_totp(params[:totp_code])
+      @user.totp_recovery_codes = Array.new(3).map{ |_| Utils.random_str(20) }
```

Comment:
good, this seems like adequate entropy to me :)

I think it would be helpful to have something more like 6 codes rather than 3. of course the normal path is to only ever use a single one, then regenerate them all, but I've been in situations where things were too urgent to do that right away, for example because I was dealing with a real-world emergency of some kind that happened to include losing access to stuff as one of its elements. In those times it was reassuring to know I wasn't going to lock myself out of the recovery process by accidentally churning through all of them.

Reply:
Should we also datestamp them, like `20250310_#{entropy}`? I can see that enabling the flow will regenerate all the existing ones, so it would be nice debugging info for the user who finds old codes, and for the admin (me) if someone tells me their code isn't working.